disk_stats.c

```diff
@@ -370,20 +383,108 @@ static void diff_disk_stats(disk_stats *new_stats)
 	}
 }
 
+static unsigned char find_or_add_device(device_stats *devices, const char *name);
+
+static void resolve_device_hierarchy(device_stats *devs, int id)
+{
+	int len, slave_id, i = 0;
+	DIR *dir;
+	struct dirent *e = NULL;
+	char slaves[255] = "/sys/block/";
+	char path[PATH_MAX];
+	char *p;
+
+	strcpy(slaves + 11, devs->values[id].name);
+	strcpy(slaves + 11 + devs->values[id].name_len, "/slaves");
+
+	if (NULL != (dir = opendir(slaves))) {
+		while (errno = 0, NULL != (e = readdir(dir)))
+			if (e->d_name[0] != '.' || (e->d_name[1]
+					&& (e->d_name[1] != '.' || e->d_name[2]))) {
+				p = NULL;
+				snprintf(path, sizeof(path), "%s/start", e->d_name);
+				if (faccessat(dirfd(dir), path, R_OK, 0) == 0 &&
+						(len = readlinkat(dirfd(dir), e->d_name, path, sizeof(path) - 1)) > 0) {
+					path[len] = '\0';
+					if ((p = strrchr(path, '/'))) {
+						*p = '\0';
+						if ((p = strrchr(path, '/')))
+							*(p++) = '\0';
+					}
+				}
+				slave_id = find_or_add_device(devs, p == NULL ? e->d_name : p);
+				devs->values[id].slaves[i++] = slave_id;
+			}
+		closedir(dir);
+	}
+	devs->values[id].slave_size = i;
+}
+
+static unsigned char find_or_add_device(device_stats *devices, const char *name)
+{
+	int i, new_id = -1;
+
+	for (i = 0; i < devices->size; ++i)
+		if (0 == strcmp(devices->values[i].name, name)) {
+			new_id = i;
+			goto resolve_hierarchy;
+		}
+
+	if (new_id == -1 && (new_id = devices->size++) >= devices->len)
+		devices->values = repalloc(devices->values, sizeof(device_stat)*(devices->len = devices->size));
+
+	memset(devices->values + new_id, 0, sizeof(device_stat));
+	devices->values[new_id].name = pstrdup(name);
```

Comment:
Since bg_mon executed in the default memory context (which is TopMemoryContext), I believe here can be a memory leak, since we never reset this context. There is a cleanup, but only for those devices that aren't in use.